File: MobileAppReactNativeHooks/src/business/dao/SimulatedDao.ts
import { Dao } from './Dao';
import { AvailableStock } from '../state/AvailableStock';
import { print, round } from '../utils/utils';

type _Stock = {
  readonly ticker: string;
  readonly name: string;
  price: number;
};

export class SimulatedDao extends Dao {

  // We simulate that the backend has this list of available stocks.
  private stocks: _Stock[] = [
    { ticker: 'IBM', name: 'International Business Machines', price: 132.64 },
    { ticker: 'AAPL', name: 'Apple', price: 183.58 },
    { ticker: 'GOOG', name: 'Alphabet', price: 126.63 },
    { ticker: 'AMZN', name: 'Amazon', price: 125.30 },
    { ticker: 'META', name: 'Meta Platforms', price: 271.39 },
    { ticker: 'INTC', name: 'Intel', price: 29.86 }
  ];

  /**
   * Read the current list of available stocks from the backend.
   * Note: this demonstrates how to get one-off information from the backend.
   */
  async readAvailableStocks(): Promise<AvailableStock[]> {

    // TODO:
    // // Wait for 0.5 second to simulate a slow network.
    // await new Promise(resolve => setTimeout(resolve, 500));

    print(`Just read ${this.stocks.length} stocks.`);

    return this.stocks.map(stock =>
      new AvailableStock(stock.ticker, stock.name, stock.price)
    );
  }

  private static _onUpdate: ((ticker: string, price: number) => void) | null = null;
  private static _priceUpdateInterval: NodeJS.Timeout | null = null;

  /**
   * Continuously get stock price updates.
   * Note: this demonstrates how to get information from websockets.
   *
   * @param onUpdate - A callback function that gets the ticker and the new
   * price of a stock.
   *
   * Example:
   * listenToStockPriceUpdates((ticker, price) => {
   *   console.log(`The new price of ${ticker} is $${price}`);
   * });
   */
  listenToStockPriceUpdates(onUpdate: (ticker: string, price: number) => void): void {

    // Simulates stock price updates, invoking the provided callback function
    // with the updated price at a regular interval. The stock price fluctuates
    // randomly within a defined range, every 250 milliseconds, selects a random
    // stock from the available stock list, simulates a price fluctuation within
    // the range of US$ 1 to US$ 1000, and invokes the onUpdate callback.

    SimulatedDao._onUpdate = onUpdate;

    // Clear previous interval.
    if (SimulatedDao._priceUpdateInterval) {
      clearInterval(SimulatedDao._priceUpdateInterval);
    }

    SimulatedDao._priceUpdateInterval = setInterval(() => {

      if (SimulatedDao._onUpdate) {
        if (this.stocks.length === 0) return;

        const randomIndex = Math.floor(Math.random() * this.stocks.length);
        const randomStock = this.stocks[randomIndex];

        // Simulates a price fluctuation.
        let newPrice = round(randomStock.price + (Math.random() * 2 - 1) / 4); // Random fluctuation

        // Limit stock price to between US$ 1 and US$ 1000.
        newPrice = Math.max(1, Math.min(newPrice, 1000));

        // Update the price in the simulated backend.
        this.stocks[randomIndex].price = newPrice;

        // Call the onUpdate callback with the new price.
        SimulatedDao._onUpdate(randomStock.ticker, newPrice);
      }
    }, 250); // Update every 250 milliseconds.
  }

  /**
   * Stop getting stock price updates.
   * Note: this demonstrates how to close websockets.
   */
  stopStockPriceUpdates(): void {

    if (SimulatedDao._priceUpdateInterval) {
      clearInterval(SimulatedDao._priceUpdateInterval);
      SimulatedDao._priceUpdateInterval = null;
    }

    SimulatedDao._onUpdate = null;
  }

  public toString(): string {
    return 'SimulatedDao';
  }
}

File: MobileAppFlutterRedux/lib/client/infra/dao/simulated_dao/simulated_dao.dart
import 'dart:async';
import 'dart:math';

import 'package:fast_immutable_collections/fast_immutable_collections.dart';
import 'package:mobile_app_flutter_redux/client/infra/run_config/run_config.dart';
import 'package:mobile_app_flutter_redux/models/available_stock.dart';

import '../dao.dart';

class SimulatedDao extends Dao {
  //
  static SimulatedDao get instance => RunConfig.instance.dao as SimulatedDao;

  SimulatedDao();

  @override
  Future<void> init() async {
    await simulatesWaiting(50);
  }

  @override
  Future<IList<AvailableStock>> readAvailableStocks() async {
    await simulatesWaiting(250);

    print('Just read ${_hardcodedStocks.length} stocks.');

    return _hardcodedStocks
        .map((stock) => AvailableStock(stock.ticker, name: stock.name, currentPrice: stock.price))
        .toIList();
  }

  Function({
    required String ticker,
    required double price,
  })? _callback;

  Timer? _priceUpdateInterval;

  @override
  Future<void> startListeningToStockPriceUpdates({required PriceUpdate callback}) async {
    _callback = callback;

    _priceUpdateInterval?.cancel();

    _priceUpdateInterval = Timer.periodic(const Duration(milliseconds: 300), (Timer t) {
      if (_callback != null) {
        final stocks = _hardcodedStocks;
        if (stocks.isEmpty) return;

        final randomIndex = Random().nextInt(stocks.length);
        final randomStock = stocks[randomIndex];

        double newPrice = (randomStock.price + (Random().nextDouble() * 2 - 1) / 4).toDouble();

        newPrice = max(1, min(newPrice, 1000));

        stocks[randomIndex].price = newPrice;

        _callback!(
          ticker: randomStock.ticker,
          price: newPrice,
        );
      }
    });
  }

  @override
  Future<void> stopListeningToStockPriceUpdates() async {
    _priceUpdateInterval?.cancel();
    _priceUpdateInterval = null;

    _callback = null;
  }
}

final List<_Stock> _hardcodedStocks = [
  _Stock('IBM', 'International Business Machines', 132.64),
  _Stock('AAPL', 'Apple', 183.58),
  _Stock('GOOG', 'Alphabet', 126.63),
  _Stock('AMZN', 'Amazon', 125.30),
  _Stock('META', 'Meta Platforms', 271.39),
  _Stock('INTC', 'Intel', 29.86),
];

class _Stock {
  final String ticker;
  final String name;
  double price;

  _Stock(this.ticker, this.name, this.price);
}

/// Simulates waiting for the demonstrating with the app.
/// For tests, will wait a very small random amount of time.
Future<void> simulatesWaiting(int millis) {
  Random _random = Random.secure();
  return (RunConfig.instance.disablePlatformChannels)
      ? Future.delayed(Duration(milliseconds: _random.nextInt(20) + 1))
      : Future.delayed(Duration(milliseconds: millis));
}
